### DISCUSSION TITLE
Looking for help getting this library to build and install

### DISCUSSION THREAD

**Discussion by jhodgson2** [gnustep/libs-base#284] [General]
Hi! I've been trying to use this library but haven't been able to do so, this is what i've achieved:

On linux I can get the library to build and run tests but a portion of the library doesn't build due to linux missing an "ng framework" but I find no references of this ng framework in the docs. Stackoverflow says these must be the apple proprietary libraries so it will absolutely not work on linux.

On macOS I am able to build but none of the test run, i get errors about duplicate method declaration of the apple stuff but i'm unable to find anyone else getting this issue, am I doing something wrong? 

I want to be able to build and have the tests pass.

Thank you!

**Comment by rfm** [comment]
You mention 'ng' as an issue.  That's part of the 'library combo' referring to the selection you made when configuring gnustep-make to decide whether you want a 'traditional' environment (building using the gcc compiler and its objc runtime) or a 'next generation' environment using the llvm/clang compiler and the matching objc runtime (which you can build/install separately).

As you 'chose' (it's the default on most systems) the traditional environment, some features (objective-c 2 language) are not available, but the library is still perfectly functional.  The automated regression tests should tell you where features are missing because your environment hasn't been set up to support them, but unless you specifically want to use a particular feature, that doesn't matter;  only actual test failures indicate a bug/issue.

There's no such thing as an 'ng framework' and stack overflow is wrong if it suggests a dependency on apple proprietory libraries.

If you actually have build errors (rather than warnings that some features are unsupported etc) you will need to look at the build logs to see what specific issue caused the build to fail.

**Comment by jhodgson2** [comment]
I was able to build and install libobjc2 but when I run ./configure on this repo it still says I'm missing ng runtime. I'm unable to determine what this means is there any documentation on how to set this up?

**Comment by rfm** [comment]
> I was able to build and install libobjc2 but when I run ./configure on this repo it still says I'm missing ng runtime.
No offence intended, but it doesn't say that.  To get help you should say exactly what you did (paste the exact command into your email) and exactly what the log/error messages were (eg attach the log).  The log/warning/error output is intended to let you know where/how things failed  and (ideally) make suggestions that should help you fix things,  so paraphrasing that information hides what it was intended to say behind your own interpretation and makes it hard for anyone to know what your problem actually is.

But to make a guess, you should reconfigure/reinstall gnustep-make using the correct configuration options.
The normal process (traditional objc rather than objc-2) on a clean system, and assuming you don't have root access so you want to install in a subdirectory of your home directory (/home/me/gnustep), is:
1. In the gnustep-make directory do './configure --prefix=/home/me/gnustep; make install'
2. Source the installed GNUstep.sh file (eg '. /home/me/gnustep/share/GNUstep/Makefiles/GNUstep.sh' to set up your environment
3. In the gnustep-base directory, do './configure; make; make install'
Finished.
If you are installing in the standard location, you would leave the --prefix=... part out when you configured, and you would need to be root at the point where you type 'make install'.


Now, if you want to use libobjc-2 the process is a bit more convoluted.  After (2), when your environment is set up, you build/install libobjc2 (which may be tricky, but should use the information set in (2) to decide the correct installation location), and you go back and repeat (1) and (2) again, but this time use the --with-library-combo=ng-gnu-gnu option configuring gnustep-make.  When you get to (3) the build will work, unless you didn't succeed in installing libobjc2 correctly, in which case the configure output will most likely end with something like this:
Check to make sure you have the Objective-C runtime library and its headers installed.
configure: error: Could not find Objective-C headers

That means what it says;  so if your libobjc2 is installed in a non-standard location you need to go back to (1) and reconfigure with environment variables set to point to the location where you have installed it.
eg. if the variable P contains the path to your libobjc2 installation it might be
  LDFLAGS=-L${P}/lib \
  CPPFLAGS=-I${P}/include \
  CFLAGS=-I${P}/include \
  OBJCFLAGS=-I${P}/include \
  ./configure
